MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.c

```diff
@@ -1408,6 +1408,286 @@ SmbiosTableConstruction (
+               (UINT8 *) SmbiosTable + OFFSET_OF (SMBIOS_TABLE_ENTRY_POINT, IntermediateAnchorString),
+               SmbiosTable->EntryPointLength - OFFSET_OF (SMBIOS_TABLE_ENTRY_POINT, IntermediateAnchorString)
+               );
+  return (BOOLEAN) (Checksum == 0);
+}
+
+/**
+  Validates a SMBIOS 3.0 table entry point.
+
+  @param  SmbiosTable   The SMBIOS_TABLE_3_0_ENTRY_POINT to validate.
+
+  @retval TRUE           SMBIOS table entry point is valid.
+  @retval FALSE          SMBIOS table entry point is malformed.
+
+**/
+STATIC
+BOOLEAN
+IsValidSmbios30Table (
+  IN SMBIOS_TABLE_3_0_ENTRY_POINT  *SmbiosTable
+  )
+{
+  UINT8 Checksum;
+
+  if (CompareMem (SmbiosTable->AnchorString, "_SM3_", 5) != 0) {
+    return FALSE;
+  }
+  if (SmbiosTable->EntryPointLength < sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT)) {
+    return FALSE;
+  }
+  if (SmbiosTable->MajorVersion < 3) {
+    return FALSE;
+  }
+
+  //
+  // The whole struct check sum should be zero
+  //
+  Checksum = CalculateSum8 (
+               (UINT8 *) SmbiosTable,
+               SmbiosTable->EntryPointLength
+               );
+  if (Checksum != 0) {
+    return FALSE;
+  }
+  return TRUE;
+}
+
+/**
+  Parse an existing SMBIOS table and insert it using SmbiosAdd.
+
+  @param  ImageHandle           The EFI_HANDLE to this driver.
+  @param  Smbios                The SMBIOS table to parse.
+  @param  Length                The length of the SMBIOS table.
+
+  @retval EFI_SUCCESS           SMBIOS table was parsed and installed.
+  @retval EFI_OUT_OF_RESOURCES  Record was not added due to lack of system resources.
+  @retval EFI_INVALID_PARAMETER Smbios is not a correct smbios table
+
+**/
+STATIC
+EFI_STATUS
+ParseAndAddExistingSmbiosTable (
+  IN EFI_HANDLE                    ImageHandle,
+  IN SMBIOS_STRUCTURE_POINTER      Smbios,
+  IN UINTN                         Length
+  )
+{
+  EFI_STATUS                    Status;
+  CHAR8                         *String;
+  EFI_SMBIOS_HANDLE             SmbiosHandle;
+  SMBIOS_STRUCTURE_POINTER      SmbiosEnd;
+
+  SmbiosEnd.Raw = Smbios.Raw + Length;
+
+  if (Smbios.Raw >= SmbiosEnd.Raw || Smbios.Raw == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  do {
+    //
+    // Make sure not to access memory beyond SmbiosEnd
+    //
+    if (Smbios.Raw + sizeof (SMBIOS_STRUCTURE) > SmbiosEnd.Raw ||
+      Smbios.Raw + sizeof (SMBIOS_STRUCTURE) < Smbios.Raw) {
+      return EFI_INVALID_PARAMETER;
+    }
+    //
+    // Check for end marker
+    //
+    if (Smbios.Hdr->Type == SMBIOS_TYPE_END_OF_TABLE) {
+      break;
+    }
+    //
+    // Make sure not to access memory beyond SmbiosEnd
+    // Each structure shall be terminated by a double-null (0000h).
+    //
+    if (Smbios.Raw + Smbios.Hdr->Length + 2 * sizeof (UINT8) > SmbiosEnd.Raw ||
+      Smbios.Raw + Smbios.Hdr->Length + 2 * sizeof (UINT8) < Smbios.Raw) {
+      return EFI_INVALID_PARAMETER;
+    }
+    //
+    // Install the table
+    //
+    SmbiosHandle = SMBIOS_HANDLE_PI_RESERVED;

```

Comment:
Use "SmbiosHandle = Smbios.Hdr->Handle.".
This is to handle the case the "Handle" was assigned properly in bootloader phase.